### ISSUE TITLE
[M1] Discuss ways to do the mount on an EFS

### ISSUE THREAD

**Issue by imaffe** [imaffe/external-workspace-manager-plugin#3]
There are some problems for the mounting procedure

1. The EFS mounting tutorial says we can : login to EC2 instance via SSH -> download EFS mount helper -> execute `mount <MOUNT_TARGET>` and this will do.
2. But I thought there might be API to do this, hence not using bash commands.
3. Then I didn't find an API in AWS Java SDK API documentation for us to execute mount on and EC2 instance like `MountEfsRequest()`
4. Does anyone has any idea how do we mount an EFS to a selected EC2

**Comment by alexsomai** [comment]
I don't know about AWS API in terms of Java SDK. But if you say there isn't any other option, probably our last choice is to invoke bash commands within Java.

**Comment by martinda** [comment]
I am not sure, but it looks like the [createMountTarget](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/elasticfilesystem/AbstractAmazonElasticFileSystem.html#createMountTarget-com.amazonaws.services.elasticfilesystem.model.CreateMountTargetRequest-) description talks about mounting:

> Creates a mount target for a file system. You can then mount the file system on EC2 instances by using the mount target.

**Comment by imaffe** [comment]
> I am not sure, but it looks like the [createMountTarget](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/elasticfilesystem/AbstractAmazonElasticFileSystem.html#createMountTarget-com.amazonaws.services.elasticfilesystem.model.CreateMountTargetRequest-) description talks about mounting:
> 
> > Creates a mount target for a file system. You can then mount the file system on EC2 instances by using the mount target.

Mount target is more like a hook, e.g it has target ID 6fde3c, then the mount would be `mount fs-6fde3c`, it does not execute the mounting, just create a hook.

EDIT : the above definition is wrong , a mount target is more like a firewall, it defines security groups for each AZ. Looks like only if an AZ has mount target, then the EC2 in it can mount the file system and others are not able to mount the NFS if no mount target specified in that AZ. Anyway after all these search I think there is no Java API for calling simple mounting, as the mount operation is more like Linux responsibility and AWS just offered some utils to get things done quickly. It is not part of the AWS CLI, so might possibly not part of AWS SDK as well. I decide to do it using scripts, though it is very unnatural ...

**Comment by imaffe** [comment]
> I don't know about AWS API in terms of Java SDK. But if you say there isn't any other option, probably our last choice is to invoke bash commands within Java.

Yeah, that would be the last choice. But I know there are IaaC software like Terraform which can configure EFS mounting for EC2 instances, there might be other choices. Anyway I'm still investigating, and this might need @hynespm 's help ~

**Comment by hynespm** [comment]
Hi @imaffe 

I have worked a lot with EFS and its setup but using Cloudformation. 

If you wish to establish an EFS, the steps you take are these

- create the file system (across multiple AZ's if you like)
- create the mount targets for each EFS (they act like firewalls)
- Establish a link between the EC2 instances SG and that of the EFS meaning that the nfs protocol can work between the two entities

That covers the infrastructure element of the connection etc. But the actual mounting is done via commands at the OS level.

So in my view, first understand the steps to link the pieces of infrastructure. Read documentation like this

- https://docs.aws.amazon.com/efs/latest/ug/mounting-fs.html

Then try and find the corresponding API call done within the Java SDK
- https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/elasticfilesystem/AmazonElasticFileSystem.html

The execution of OS level scripts can probably also be done, but first master the linking of the infrastructure. If all is done correctly, the mount command (at least for Linux) can be straightforward. 

Also, bear in mind that OS should, I believe have nfs-utils package installed. But worry about that part later.

Hope this helps.

**Comment by imaffe** [comment]
Hi @hynespm,
Recently I'm working with Terraform like Cloudformation and I read all the documents you mentioned. Got some questions 
1. I thought user manage EFS creation and destroy and not by Jenkins, since it ask user to assign an appropriate IAM role to Jenkins, and that is hard. Also, if we do on-demand creating/destroying for each job/build, I'm not sure about the pricing policy. I viewed the pricing for EFS, like there is no second-level pricing policy, mostly month and hour(maybe), so I wonder if it is natural to create EFS on-demand. And we can discuss  in the meeting whether we should leave creating EFS to user .
2. For the corresponding API calls, I searched 10 days ago and yesterday, for me I didn't find an API call on EC2 instance to mount something. I think there is no API for mounting because : 1 . in terraform there is no configuration option for 'mounting', so there might not be one 2. in the AWS documentation, it never mentioned mounting using API, but for other services like sg configuration, they put it on the documentation. 3. They say we have to download an efs-mount-helper program to manage the mount, and that might not be part of the SDK. according to the reasons above, I'd say we can only use scripts to do the mount/unmount. But in case I missed something, you can search the API in a more detailed way as I am not confident about myself....

**Comment by imaffe** [comment]
So basically we need to figure out what's jenkins' responsibility ,  we have some options : 
1. whether EWM is responsible for creating a new EC2 or not.
2 .whether EWM pre-mount the EFS or on-demand mount the EFS?
3. whether EWM is responsible for creating a EFS on-demand ? or it can manage the EFS for a time period like 1 or 2 months ?
when we have the above questions answered, we can go on to decide : 
1. what information should user give to jenkins (iam role accessKey and accessId?)
I'm now comfortable with AWS Java SDK, (read codes in ec2-plugin and aws documents), and I'd say the access control is the most difficult part, as we need to check whether user gave us just enough rights.

**Comment by imaffe** [comment]
@martinda @alexsomai @hynespm I will post the questions we are gonna discuss this Sunday here, and the above are some of them.

**Comment by martinda** [comment]
I would prefer if these were asked on the dev mailing list. The audience would be much broader with more chances of getting more answers.

**Comment by imaffe** [comment]
> I would prefer if these were asked on the dev mailing list. The audience would be much broader with more chances of getting more answers.

These are more like plugin specific decisions, so others might know the context of it. I'd prefer we discuss them first and if we disagree with each other, we can ask for other developers to give us hints.

**Comment by martinda** [comment]
@imaffe There is a special interest group for cloud features in Jenkins, it's called Cloud Native. This is the [page](https://jenkins.io/sigs/cloud-native/) and this is the [chat room](https://gitter.im/jenkinsci/cloud-native-sig). I suggest you ping them with questions you have such as the ones here.
